### PR TITLE
feat: add Asana ticket detection to ticket compliance check

### LINKED ISSUE #2002: Add Support for Asana Ticket Provider
### Feature request

Extend the existing infrastructure of ticket fetching and compliance in the review tool to the Asana Ticket Provider

### Motivation

Asana is a software-as-a-service platform designed for team collaboration and work management. Teams can create projects, assign tasks, set deadlines, and communicate directly within Asana. It also includes reporting tools, file attachments, calendars, and goal tracking.

https://en.wikipedia.org/wiki/Asana,_Inc.

### PR DESCRIPTION
**Pull request by Oxygen56**
## What

Fixes #2002 — adds authenticated Asana task fetching to the ticket compliance check.

## Problem

The ticket compliance check could use GitHub and Azure DevOps ticket content, but Asana references did not provide real task context for compliance analysis.

## Solution

- Detect full Asana task URLs in both legacy and current permalink formats.
- Fetch each referenced task through `GET /api/1.0/tasks/{task_gid}` using a configured Bearer token.
- Keep the token-bearing request target fixed to Asana's official API as a security boundary.
- Include the task name, notes, tags, and canonical permalink in ticket context.
- Deduplicate references by task GID while preserving their first-seen order.
- Bound Asana fetch attempts and isolate invalid, missing, inaccessible, or failed tasks so one bad reference cannot abort the batch.
- Preserve native provider behavior: GitHub keeps its existing three-issue limit, Azure keeps all linked work items, and each path appends at most three Asana tasks.

## Configuration

Configure an Asana personal access token in `.secrets.toml`:

```toml
[asana]
api_token = "YOUR_PERSONAL_ACCESS_TOKEN"
```

Environment-based deployments can use `ASANA__API_TOKEN`. The default per-task request timeout is 10 seconds, can be changed with `asana.request_timeout`, and is capped at 60 seconds.

## Local verification

- Focused ticket-compliance and extraction tests: 64 passed.
- Full unit suite with coverage: 1783 passed, 1 skipped, 1 xfailed.
- Ruff, flake8, isort, pre-commit hooks, and diff whitespace checks passed.
- MkDocs site build passed.
- No live Asana credentials were used; API behavior is covered with mocked responses.

### CODE REVIEW

**Comment by naorpeled** [comment]
Hey @Oxygen56,
Thanks for working on this!

Can you please add tests to validate the new functionality? 🙏

**Comment by Oxygen56** [comment]
感谢所有 review 意见，我已在最新提交中处理了以下问题：

**Bug 修复：**
- ✅ 不安全的 `dline[remove_count:]` 去缩进 → 改为只剥离每行实际的空白字符
- ✅ Asana ticket 被截断丢弃 → 重构截断逻辑，存在 Asana ticket 时至少保留 1 个位置
- ✅ Asana 缩写大小写敏感 → 添加 `re.IGNORECASE`，现在支持 ASANA-/asana-/Asana-
- ✅ Asana 缩写过度匹配 → 添加 `\b` 单词边界

**规范/安全修复：**
- ✅ 移除未使用的 `import pytest`
- ✅ `_ASANA_TASK_*_PATTERN` 常量现在在 `find_asana_tickets()` 中使用
- ✅ 单引号改为双引号
- ✅ CodeQL: 使用 `startswith()` 替代子串匹配

**测试补充：**
- ✅ 新增 `test_shorthand_mixed_case_asana` 测试
- ✅ 新增 `test_shorthand_respects_word_boundary` 测试

File: pr_agent/tools/ticket_pr_compliance_check.py [hunk 1]
```diff
@@ -133,6 +171,19 @@
             if tickets:
 
                 for ticket in tickets:
+                    # Skip Asana URLs — these are external references,
+                    # included for visibility but cannot be fetched via GitHub API.
+                    if "app.asana.com" in ticket:
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py]: 已修复：使用 `ticket.startswith("https://app.asana.com/")` 替代子串匹配。

File: pr_agent/tools/ticket_pr_compliance_check.py [hunk 2]
```diff
@@ -35,6 +35,37 @@
     return list(tickets)
 
 
+# Compiled Asana task patterns
+_ASANA_TASK_URL_PATTERN = re.compile(
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py:42]: 已修复：`_ASANA_TASK_URL_PATTERN` 现在在 `find_asana_tickets()` 中通过 `.finditer()` 使用。

File: pr_agent/tools/ticket_pr_compliance_check.py [hunk 3]
```diff
@@ -35,6 +35,37 @@
...
+    r'https://app\.asana\.com/0/(\d+)/(\d+)'
+)
+_ASANA_TASK_SHORT_PATTERN = re.compile(
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py]: 已修复：`_ASANA_TASK_SHORT_PATTERN` 现在在 `find_asana_tickets()` 中使用，并添加了 `re.IGNORECASE` 和 `\b` 边界。

File: pr_agent/tools/ticket_pr_compliance_check.py [hunk 4]
```diff
@@ -35,6 +35,37 @@ def find_jira_tickets(text):
...
+    Returns:
+        A list of Asana task URLs.
+    """
+    tickets = set()
+    for match in re.finditer(r'https://app\.asana\.com/0/(\d+)/(\d+)', text):
+        tickets.add(match.group(0))
+    for match in re.finditer(r'(?:^|[^A-Za-z0-9])(?:ASANA|asana)[- ]?(\d{12,20})', text):
+        task_id = match.group(1)
+        tickets.add(f"https://app.asana.com/0/0/{task_id}")
+    return sorted(tickets)
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py]: 已修复：新增了 `test_shorthand_mixed_case_asana` 和 `test_shorthand_respects_word_boundary` 测试用例。

File: pr_agent/tools/ticket_pr_compliance_check.py [hunk 5]
```diff
@@ -123,6 +154,13 @@ async def extract_tickets(git_provider):
...
+
+            # Also detect Asana ticket references in the PR description
+            asana_tickets = find_asana_tickets(user_description)
+            for link in asana_tickets:
+                if link not in seen:
+                    seen.add(link)
+                    merged.append(link)
             if len(merged) > 3:
                 get_logger().info(f"Too many tickets (description + branch): {len(merged)}")
                 tickets = merged[:3]
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py]: 已修复：重构了截断逻辑，当存在 Asana ticket 时至少为其保留 1 个位置（`github_like[:2] + asana_links[:1]`，再填充剩余位置）。

File: pr_agent/tools/pr_code_suggestions.py [hunk 6]
```diff
@@ -609,10 +609,43 @@ def dedent_code(self, relevant_file, relevant_lines_start, new_code_snippet):
...
+                    # Remove exactly -delta_spaces leading spaces from each
+                    # non-empty line.  textwrap.dedent() strips the *common*
+                    # indent which may remove too much when lines have
+                    # varying indentation levels (Qodo bug #3).
+                    remove_count = -delta_spaces
+                    dedented_lines = []
+                    for dline in new_code_snippet.split('\n'):
+                        if dline.strip():
+                            dedented_lines.append(dline[remove_count:])
+                        else:
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/pr_code_suggestions.py]: 已修复：改为按行只剥离实际空白字符 `min(remove_count, leading)`，不再盲目切片，避免在缩进不一致时删除实际代码。

File: tests/unittest/test_ticket_compliance.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
+"""
+Unit tests for Asana ticket detection in ticket_pr_compliance_check.py.
+
+Tests cover:
+- Full Asana URL detection
+- Shorthand ASANA- prefix detection
+- Edge cases (mixed content, no tickets, duplicates)
+"""
+import pytest
+from pr_agent.tools.ticket_pr_compliance_check import find_asana_tickets
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply tests/unittest/test_ticket_compliance.py]: 已修复：移除了未使用的 `import pytest`。

File: pr_agent/tools/pr_code_suggestions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/pr_code_suggestions.py]: 已修复：改为按行只剥离实际空白字符 `min(remove_count, leading)`，不再盲目切片。

File: pr_agent/tools/ticket_pr_compliance_check.py [hunk 8]
```diff
@@ -133,6 +171,17 @@ async def extract_tickets(git_provider):
...
+                    # Skip Asana URLs — these are external references,
+                    # included for visibility but cannot be fetched via GitHub API.
+                    if "app.asana.com" in ticket:
+                        tickets_content.append({
+                            "title": f"Asana Task: {ticket}",
+                            "url": ticket,
+                            "body": ("Asana task referenced in PR description. "
+                                     "Fetch task details from Asana for full context."),
+                        })
+                        continue
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py]: 此 PR 的目标是 **检测** Asana ticket 引用（对应 issue #2002 的需求）。完整的 Asana API 集成（包括获取 ticket 详情）属于后续独立 feature，需要 Asana API token 等额外配置。当前实现将 Asana ticket 加入 `related_tickets` 列表，确保它们在 compliance check 中可见。

**Comment by Oxygen56** [comment]
Hey @naorpeled, thanks for the review! 

Tests have been added — there are now 12 test cases in `tests/unittest/test_ticket_compliance.py` covering:

- Full Asana URL detection (`https://app.asana.com/0/...`)
- Shorthand format (`ASANA-123456789012`)
- Case-insensitive matching (`asana-`, `Asana-`)
- Word boundary enforcement (no false matches on `NOTASANA-...`)
- Deduplication of identical tickets
- Multiple tickets in mixed content
- Empty input / no-ticket text
- GitHub URL exclusion
- Sorted output

The test file was added in commit `a0c62b4` and extended with two additional edge cases in the latest push (`3ce5450`). Let me know if anything else is needed!

File: pr_agent/tools/pr_code_suggestions.py [hunk 9]
```diff
@@ -609,10 +609,44 @@ def dedent_code(self, relevant_file, relevant_lines_start, new_code_snippet):
                 original_initial_spaces = len(original_initial_line) - len(original_initial_line.lstrip()) # lstrip works both for spaces and tabs
                 suggested_initial_spaces = len(suggested_initial_line) - len(suggested_initial_line.lstrip())
                 delta_spaces = original_initial_spaces - suggested_initial_spaces
+                # Detect indentation character from original line
+                indent_char = '\t' if original_initial_line.startswith('\t') else ' '
```
Comment (naorpeled) [inline pr_agent/tools/pr_code_suggestions.py]: I think it's better to revert this as it's unrelated to the Asana  detection, wdyt?

**Comment by dellch** [comment]
There are docs (`docs/docs/core-abilities/fetching_ticket_context.md`) that perhaps should be updated as well?

File: pr_agent/tools/pr_code_suggestions.py (hunk 9 above)
Comment (Oxygen56) [reply pr_agent/tools/pr_code_suggestions.py]: Done — reverted the unrelated indentation changes in `pr_code_suggestions.py`. The file is now back to its original state from `main`. Only the Asana-related changes remain in `ticket_pr_compliance_check.py` and `tests/unittest/test_ticket_compliance.py`.

**Comment by Oxygen56** [comment]
Updated `docs/docs/core-abilities/fetching_ticket_context.md` with an Asana Integration section covering supported reference formats (full URLs and `ASANA-123456789012` shorthand), how to link a PR to an Asana task, and a note that Asana task content is not fetched via API (reference-only for compliance checking).

**Comment by Oxygen56** [comment]
Tests for the Asana ticket detection are already included in this PR — see `tests/unittest/test_ticket_compliance.py`, which was added with 96 lines covering both full Asana URL and shorthand format detection.

File: pr_agent/tools/ticket_pr_compliance_check.py [hunk 10]
```diff
@@ -35,6 +35,38 @@ def find_jira_tickets(text):
     return list(tickets)
 
 
+# Compiled Asana task patterns
+_ASANA_TASK_URL_PATTERN = re.compile(
+    r'https://app\.asana\.com/0/(\d+)/(\d+)'
+)
+_ASANA_TASK_SHORT_PATTERN = re.compile(
+    r'(?:^|[^A-Za-z0-9])(?:ASANA|asana)[- ]?(\d{12,20})',
```
Comment (naorpeled) [inline pr_agent/tools/ticket_pr_compliance_check.py]: Are you sure this is valid?
Afaik, Asana doesn't have a specific format for ids

Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py]: Good catch — you're right, Asana doesn't have a native shorthand format like Jira's `PROJ-123`. Asana task IDs are just numeric strings (e.g., `1234567890123456`).

The `ASANA-` prefix was introduced here as a pragmatic convention so PR authors can reference Asana tasks without pasting full `https://app.asana.com/...` URLs. It's not an Asana-native format.

I can see two reasonable paths:

1. **Drop the shorthand entirely** — only support full Asana URLs (`https://app.asana.com/0/{project_id}/{task_id}`). Simpler, no invented convention.

2. **Keep it** — treat it as a PR-Agent-specific convenience notation, not an Asana feature. Helps users who want to type a quick reference.

What would you prefer? Happy to go either way.

Comment (naorpeled) [reply pr_agent/tools/ticket_pr_compliance_check.py]: I think 1.

Comment (Oxygen56) [reply pr_agent/tools/ticket_pr_compliance_check.py]: Done — dropped the `ASANA-` shorthand format entirely. Now only full Asana URLs (`https://app.asana.com/0/{project_id}/{task_id}`) are supported. Updated the docs and tests accordingly.

**Comment by naorpeled** [comment]
Hey @Oxygen56,
this overall LGTM, can you please address the last comments by Qodo?

Thanks!

**Comment by Oxygen56** [comment]
Thanks for the nudge. I addressed the remaining Qodo items in the latest push:

- The unused `pytest` import is no longer present in `tests/unittest/test_ticket_compliance.py`.
- Added integration coverage for `extract_tickets()` so Asana URLs are validated in the merged ticket-content path, not only in the standalone parser.
- Added coverage for the capped-ticket path to ensure an Asana URL is still retained when the related ticket list exceeds the 3-ticket limit.

Validation run locally:
- `.venv/bin/python -m pytest tests/unittest/test_ticket_compliance.py -q` -> 10 passed
- `.venv/bin/ruff check tests/unittest/test_ticket_compliance.py` -> passed
- `.venv/bin/ruff check --select F401 tests/unittest/test_ticket_compliance.py` -> passed
- `git diff --check` -> passed

**Comment by Oxygen56** [comment]
Thanks, I pushed bc043c7 to address the actionable new Qodo items:

- `find_asana_tickets()` now returns `[]` for `None`, non-string, and empty input, so `extract_tickets()` no longer reaches regex matching with invalid text.
- The 3-ticket truncation path now backfills from available Asana links while still preserving at least one Asana reference when present.
- Added tests for `None` input and the all-Asana truncation/backfill case.

Validation:
- `.venv/bin/python -m pytest tests/unittest/test_ticket_compliance.py -q` -> 12 passed
- `.venv/bin/ruff check tests/unittest/test_ticket_compliance.py` -> passed
- `.venv/bin/python -m py_compile pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `git diff --check` -> passed

On the broader Asana fetching note: this PR currently keeps Asana references as detection/reference-only, matching the docs added here. Implementing full Asana API fetching would require new provider config/auth handling, so I am keeping that out of scope unless you want this PR expanded in that direction.

**Comment by Oxygen56** [comment]
Pushed cf17141 for the latest Qodo pass:

- Removed the inner broad catch around `get_user_description()` so failures are no longer silently converted to an empty description by the new Asana path.
- Added a shared cap for provider-agnostic Asana ticket entries so non-GitHub fallback and Azure paths cannot add unbounded Asana references.
- Azure now preserves one slot for an Asana reference when the work-item list is already at the 3-ticket cap.
- Added tests for non-GitHub Asana capping and Azure + Asana capping.

Validation:
- `.venv/bin/python -m pytest tests/unittest/test_ticket_compliance.py -q` -> 15 passed
- `.venv/bin/ruff check tests/unittest/test_ticket_compliance.py` -> passed
- `.venv/bin/python -m py_compile pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `git diff --check` -> passed

The remaining full Asana API fetching/provider-auth item is still a scope question rather than a small follow-up fix.

**Comment by Oxygen56** [comment]
Pushed 1c1a693 for the latest Qodo pass:

- `max_related_tickets` is now configurable via `pr_reviewer.max_related_tickets` with a default of 3.
- The Asana URL regex now follows the project double-quote style.
- Azure DevOps linked work items are now capped even when no Asana references are present, while still reserving one slot for Asana when applicable.
- Added tests for configurable ticket caps and Azure-only capping.

Validation:
- `.venv/bin/python -m pytest tests/unittest/test_ticket_compliance.py -q` -> 17 passed
- `.venv/bin/ruff check pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `.venv/bin/python -m py_compile pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `git diff --check` -> passed

A couple of remaining Qodo entries look stale or scope-level: non-GitHub Asana detection is already provider-agnostic and covered by `test_extract_tickets_includes_asana_for_non_github_provider`; the full Asana API fetching/provider-auth item would require a broader config/auth provider implementation beyond this reference-detection PR.

**Comment by Oxygen56** [comment]
Follow-up pushed as c928e6c for the same `extract_tickets()` return-path item:

- The success path now returns `tickets_content` unconditionally, so no-ticket cases return `[]` directly.
- The exception path still returns `[]`.

Validation remains green:
- `.venv/bin/python -m pytest tests/unittest/test_ticket_compliance.py -q` -> 19 passed
- `.venv/bin/ruff check pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `.venv/bin/python -m py_compile pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `git diff --check` -> passed

**Comment by Oxygen56** [comment]
Pushed 557b28c for the new Qodo `Asana slots underfilled` item:

- The GitHub path now keeps the full Asana candidate list available for the shared append helper.
- If a GitHub issue fetch is skipped after truncation, Asana references can still backfill the remaining slots up to `max_related_tickets`.
- Added regression coverage for a GitHub issue fetch failure followed by Asana backfill.

Validation:
- `.venv/bin/python -m pytest tests/unittest/test_ticket_compliance.py -q` -> 20 passed
- `.venv/bin/ruff check pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `.venv/bin/python -m py_compile pr_agent/tools/ticket_pr_compliance_check.py tests/unittest/test_ticket_compliance.py` -> passed
- `git diff --check` -> passed

**Comment by IsmaelMartinez** [comment]
Thanks for building this, and apologies for the long wait.

I ran `tests/unittest/test_ticket_compliance.py` at `557b28c6` and all 20 pass, and the wiring into `extract_tickets` reads correctly. Two things before it can land.

It conflicts with `main` and needs a rebase. And `_make_asana_ticket_content` returns a fixed body telling the reader to fetch the task details, which nothing does, so the check runs against a placeholder rather than the task. Worth saying plainly, because #2002 asked to extend ticket fetching.

Also the description promises an `ASANA-123456789012` shorthand that the diff does not implement.

Can you apply those and then I can do a final pass and review?

**Comment by Oxygen56** [comment]
Thanks for the detailed review. I updated the branch onto current `main` and replaced the placeholder Asana content with authenticated task fetching through `GET /api/1.0/tasks/{task_gid}`.

The implementation now:

- maps task name, notes, tags, and the canonical permalink into ticket context;
- supports both legacy Asana links and current V1 permalinks;
- deduplicates by task GID and preserves first-seen order;
- bounds API attempts and skips missing, inaccessible, or failed tasks;
- preserves the existing GitHub and Azure DevOps behavior.

I also updated the documentation and removed the stale `ASANA-...` shorthand claim from the PR description.

Local verification:

- focused ticket tests: 63 passed;
- full unit suite with coverage: 1782 passed, 1 skipped, 1 xfailed;
- Ruff, flake8, isort, pre-commit hooks, diff checks, and the MkDocs build passed.

No live Asana credentials were used; API behavior is covered with mocked responses.

Could you please take the final pass when convenient?

**Comment by IsmaelMartinez** [comment]
Thanks for turning that around so fast, and for going further than I asked.

Verified at `d3c8eb5f`: the token only ever goes in a Bearer header and aiohttp drops it on cross-origin redirect, the timeout is wired to the session and capped, `task_gid` is always digits so there is no SSRF surface, 34 mocked tests, suite 1782 pass.

One thing worth a look: `reserved_asana_slots` at L352-353 takes a slot out of `MAX_RELATED_TICKETS`, so a PR referencing both an Asana task and three GitHub issues now silently drops the third GitHub issue from review context. I take it the reserve is deliberate, so mainly flagging that it isn't documented. The Azure path at L468 has the opposite problem, `extend` with no cap at all.

**Comment by Oxygen56** [comment]
Thanks for flagging both limit issues. I removed the shared-slot behavior and made the limits provider-specific so Asana cannot displace native tickets:

- GitHub keeps its existing three-issue budget, then appends up to three Asana tasks.
- Azure DevOps keeps all existing linked work items, then appends up to three Asana tasks.
- Other providers can contribute up to three Asana tasks.

The ordering and limits are now documented and covered by regressions for 3 GitHub + 1 Asana and 4 Azure + 4 Asana.

I also isolated malformed Asana references so one bad entry cannot abort the remaining batch, and replaced the duplicated cap literals with separate GitHub and Asana constants. The Asana API endpoint remains fixed intentionally: it is the security boundary that prevents the Bearer token from being sent to a configurable origin.

Verification:

- focused ticket tests: 64 passed
- full unit suite with coverage: 1783 passed, 1 skipped, 1 xfailed
- Ruff, flake8, isort/pre-commit, diff checks, and MkDocs build passed

Could you take the final pass when convenient?

File: tests/unittest/test_ticket_compliance.py [hunk 11]
```diff
@@ -0,0 +1,644 @@
...
+
+
+class _FirstTwoFailingRepo:
+    def get_issue(self, number):
+        if number in {1, 2}:
+            raise RuntimeError("issue unavailable")
+        return _Issue(number)
+
+
+class _GithubProvider(GithubProvider):
```
Comment (IsmaelMartinez) [inline tests/unittest/test_ticket_compliance.py]: CI ran here for the first time today and found two things, both in this file.

CodeQL `py/missing-call-to-init`, here and at :84. `extract_tickets` dispatches on `isinstance(git_provider, GithubProvider)` at :340, so the doubles do need to be real instances, but `test_ticket_extraction_async.py` gets that with `GithubProvider.__new__(GithubProvider)` rather than subclassing, which doesn't trip the rule.

`build-and-test`: 18 failures, all `async def functions are not natively supported`. The docker test target doesn't copy `pyproject.toml`, so `asyncio_mode = "auto"` doesn't apply there, which is why every async test file on main carries an explicit `@pytest.mark.asyncio`. Adding the marker here clears it.

Everything else checks out, including the provider-specific limits and the security properties.

Comment (Oxygen56) [reply tests/unittest/test_ticket_compliance.py]: Thanks for the clear pointers. I replaced the subclass-based test doubles with real provider instances created via `__new__`, matching the existing test pattern, and added explicit `@pytest.mark.asyncio` markers to all 18 async tests.\n\nLocal verification:\n- focused ticket tests in strict asyncio mode: 64 passed\n- full unit suite in strict asyncio mode: 1783 passed, 1 skipped, 1 xfailed\n- Ruff, flake8 (120-character project limit), pre-commit, and diff checks passed\n\nThe new GitHub Actions runs are currently waiting for maintainer approval.

**Review by IsmaelMartinez — APPROVED** [review]
That is it, thanks. Verified at `1b7dc524`: 18 of 18 async tests carry the marker and both subclasses are gone, so CodeQL is clear. Reproduced the Docker condition with no `pyproject.toml` and got 35 pass, against 18 failures before. Merged with current main locally, 1835 pass. Approving.
